### PR TITLE
tiling kwargs for pytorch model inputs

### PR DESCRIPTION
**Pull request by piotrm0**
Our NLP use case requires to tile model keyword arguments to align dimensions at doi_cut with dimensions of intervention. This implementation heuristically assumes the first model argument is batchable and every other argument that has the same first dimension as it are also batchable. The first model argument may be a keyword argument if no positional arguments are given.

Only pytorch backend is fully supported in this regard.

### CODE REVIEW

File: trulens/nn/models/pytorch.py [hunk 1]
```diff
@@ -284,7 +286,23 @@ def fprop(
...
+                tile_shape[0] = doi_resolution
+                repeat_shape = tuple(tile_shape)
+
+                if isinstance(val, np.ndarray):
+                    val = np.tile(val, repeat_shape)
+
+                elif torch.is_tensor(val):
+                    val = val.repeat(repeat_shape)
+
+                batched_model_kwargs[k] = val
```
Comment (rshih32) [inline trulens/nn/models/pytorch.py:302]: nit: let's put this in a shared function that both can call.

File: trulens/nn/models/pytorch.py [hunk 2]
```diff
@@ -269,7 +269,9 @@ def fprop(
         else:
             doi_repeated_batch_size = intervention[0].shape[0]
             batched_model_args = []
+            batched_model_kwargs = {}
 
+            # tile batches for interpolation
             for val in model_args:
                 doi_resolution = int(doi_repeated_batch_size / val.shape[0])
                 tile_shape = [1 for _ in range(len(val.shape))]
```
Comment (rshih32) [inline trulens/nn/models/pytorch.py:277]: Just a note, depending on how the memory task goes, we may want to do smarter checking and not tile everthing. We could check if doi_resolution is the expected length.
captured here: https://dev.azure.com/ailens-io/truera/_backlogs/backlog/truera%20Team/Epics/?workitem=5907

**Review by rshih32 — APPROVED** [review]

File: trulens/nn/models/pytorch.py (hunk 2 above)
Comment (piotrm0) [reply trulens/nn/models/pytorch.py:277]: Did this, the first arg is expected to be batchable, all subsequent ones are batched only if they match the first's first dim.

File: trulens/nn/models/pytorch.py (hunk 1 above)
Comment (piotrm0) [reply trulens/nn/models/pytorch.py:302]: Done.

File: trulens/nn/models/_model_base.py [hunk 3]
```diff
@@ -172,8 +242,24 @@ def fprop(
                 as the input. If `attribution_cut` is supplied, also return the
                 cut activations.
         """
+
+        # TODO: Can attribution_cut preceed doi_cut or follow to_cut in the model?
```
Comment (rshih32) [inline trulens/nn/models/_model_base.py:246]: we can remove this todo and add to documentation. There are 2 questions here:
1) attribution cut cannot preceed  doi_cut (or gradients would break since expansion is not differentiable)
2) to_cut is not specifically tied to any of the attribution progression, it just says to which layer the fprop is going to. so it can be before or after any cut, as it can be to any layer.

File: trulens/nn/models/_model_base.py [hunk 4]
```diff
@@ -172,8 +242,24 @@ def fprop(
...
+        # TODO: Use the defaults method for all backends.
+
         raise NotImplementedError
 
+    def _qoi_bprop_defaults(self, doi_cut, to_cut, attribution_cut):
+        """Defaults logic for qoi_bprop common across backends."""
```
Comment (rshih32) [inline trulens/nn/models/_model_base.py:252]: nit: same sentence fixing:

Replaces input parameters used in qoi_bprop to default values if not explicitly provided. These are common across backends.

File: trulens/nn/models/keras.py [hunk 5]
```diff
@@ -318,20 +319,56 @@ def fprop(
...
-        # Convert `intervention` to a list of inputs if it isn't already.
-        if not isinstance(intervention, DATA_CONTAINER_TYPE):
-            intervention = [intervention]
+        print("doi_tensors=", doi_tensors)
+
+        # Finally, interventions override any previously set tensors.
+        val_map.update({k: v for k, v in zip(doi_tensors, intervention)})
+        # val_map.update({_tensor(k): v for k, v in intervention.kwargs.items()})
+
+        print(val_map)
```
Comment (rshih32) [inline trulens/nn/models/keras.py:366]: remove print statements

File: trulens/nn/models/pytorch.py [hunk 6]
```diff
@@ -64,8 +68,14 @@ def __init__(
         self._layernames = list(layers.keys())
         self._tensors = list(layers.values())
 
+        if 0 == len(self._tensors):
```
Comment (rshih32) [inline trulens/nn/models/pytorch.py:71]: nit: swap the operands to be consistent with the rest of code:
len(self._tensors) == 0

File: trulens/nn/models/pytorch.py [hunk 7]
```diff
@@ -442,26 +477,32 @@ def qoi_bprop(
...
         grads_list = []
+
         for z in zs:
             z_flat = ModelWrapper._flatten(z)
             qoi_out = qoi(y)
 
+            # TODO(piotrm): this sum is a source of much bugs for me when using
+            # attributions. If one wants a specific QoI, the sum hides the bugs
+            # in the definition of that QoI. It might be better to give an error
+            # when QoI is not a scalar.
```
Comment (rshih32) [inline trulens/nn/models/pytorch.py:505]: multi qoi is used for things like average qoi. its morre efficient to do it here and have 1 gradient rather than many gradients

File: trulens/utils/typing.py [hunk 8]
```diff
@@ -5,19 +5,212 @@
...
+    # returning a new container.
+
+    @staticmethod
+    def lenses_elements(c: List[V]) -> Iterable['Lens[List[V], V]']:
+        """Lenses focusing on elements of a list."""
+
+        for i in range(len(c)):
+            yield Lens(
+                lambda l, i=i: l[i],
+                lambda l, v, i=i: IndexableUtils.with_(l, i, v))
```
Comment (rshih32) [inline trulens/utils/typing.py:107]: why do we need this class?

**Review by rshih32 — CHANGES_REQUESTED** [review]

File: tests/pytorch/unit/model_wrapper_test.py [hunk 9]
```diff
@@ -51,6 +51,43 @@ def forward(this, x):
...
+            def forward(this, X, Coeffs, divisor, **kwargs):
+                # Hoping kwargs forces the remaining arguments to be passed only by kwargs.
+
+                Degree = kwargs['Degree']
+                offset = kwargs['offset']
+
+                layer1 = this.layer1(X)
+
+                print("X=", X)
+                print("Coeffs=", Coeffs)
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply tests/pytorch/unit/model_wrapper_test.py:76]: Yes, they are to be batched (non-capitals are parameters to not be batched).

File: trulens/nn/models/pytorch.py (hunk 6 above)
Comment (piotrm0) [reply trulens/nn/models/pytorch.py:71]: I'd suggest to swap all others to "0 == ".

File: trulens/utils/typing.py (hunk 8 above)
Comment (piotrm0) [reply trulens/utils/typing.py:107]: For unifying access to different type of args in ModelInputs .

File: trulens/nn/models/_model_base.py [hunk 10]
```diff
@@ -132,22 +184,40 @@ def fprop(
...
-        `intervention`, and ending at `to_cut`.
+        Several cuts are parameters to this method. These designate how
+        information is propagated, intervened on, and returned by this method.
+        Models are evaluated starting from model_args, model_kwargs:
+
+            model_inputs -> doi_cut -> attribution_cut -> to_cut
+
+        Tensors for attribution_cut and to_cut are returned. However, the
+        returned values are batched according to the batching structure hinted
+        to by intervention. That is, if we have dimensions:
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:195]: You mean doi_cut?

File: trulens/nn/models/_model_base.py [hunk 11]
```diff
@@ -119,6 +123,54 @@ def __call__(self, x):
...
+                    "intervention needs to be given for DoI cuts that are not InputCut"
+                )
+            else:
+                # Using as_args here as sometimes interventions are passed in as single tensors but args needs a list.
+                intervention = ModelInputs(as_args(intervention), {})
+
+        return doi_cut, to_cut, intervention, model_inputs
+
     @abstractmethod
     def fprop(
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:175]: Unsure what propagate typing meas.

File: trulens/nn/models/pytorch.py (hunk 6 above)
Comment (rshih32) [reply trulens/nn/models/pytorch.py:71]: Even our inequalities are written in a way variable > constant. Not sure if there's a good reason to swap them all

File: trulens/nn/models/_model_base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rshih32) [reply trulens/nn/models/_model_base.py:175]: this fprop function needs typing.

File: trulens/nn/models/_model_base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rshih32) [reply trulens/nn/models/_model_base.py:195]: actually we can remove the cut mention, I think I didn't mean either
lets do 

However, the returned values are batched according to the batching structure hinted to by intervention, which could be larger due to expanding by the doi.

File: trulens/nn/models/_model_base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:175]: Done.

File: trulens/nn/models/_model_base.py [hunk 12]
```diff
@@ -119,6 +123,54 @@ def __call__(self, x):
...
+            doi_cut = InputCut()
+        if to_cut is None:
+            to_cut = OutputCut()
+
+        if isinstance(doi_cut, InputCut):
+            # For input cuts, produce a ModelInputs container for the intervention and model inputs.
+            if intervention is not None:
+                tru_logger.warn(
+                    "intervention for InputCut DoI specified; this is ambiguous with model_args, model_kwargs"
+                )
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:144]: Done.

File: trulens/nn/models/_model_base.py [hunk 13]
```diff
@@ -119,6 +123,54 @@ def __call__(self, x):
         """
         return self.fprop(x)[0]
 
+    def _fprop_defaults(
+            self, model_args, model_kwargs, doi_cut, to_cut, intervention):
+        """Defaults logic for fprop common across backends. Also some warnings and errors."""
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:128]: Done.

File: trulens/nn/models/_model_base.py [hunk 14]
```diff
@@ -119,6 +123,54 @@ def __call__(self, x):
         """
         return self.fprop(x)[0]
 
+    def _fprop_defaults(
+            self, model_args, model_kwargs, doi_cut, to_cut, intervention):
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:127]: Done.

File: trulens/nn/models/_model_base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:195]: Rephrased/elaborated.

File: trulens/nn/models/_model_base.py (hunk 3 above)
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:246]: Moving the documentation about these restrictions to bprop as I don't think they matter for fprop; fprop doesn't take any gradients

Comment (piotrm0) [reply trulens/nn/models/_model_base.py:246]: Also I'm unsure about your point (2), looks like the gradients done in bprop are to_cut w.r.t. attribution_cut so that implies an order regarding these two which would produce errors were it not to hold.

On the other hand, if attribution was before doi_cut, the evaluation might run but have unexpected results.

File: trulens/nn/models/_model_base.py (hunk 4 above)
Comment (piotrm0) [reply trulens/nn/models/_model_base.py:252]: Done.

File: trulens/nn/distributions.py [hunk 15]
```diff
@@ -283,7 +295,7 @@ def __init__(self, var: float, resolution: int, cut: Cut = None):
         self._var = var
         self._resolution = resolution
 
-    def __call__(self, z: ArrayLike) -> List[ArrayLike]:
+    def __call__(self, z: ArgsLike) -> List[ArgsLike]:
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/distributions.py:298]: Done.

File: trulens/nn/distributions.py [hunk 16]
```diff
@@ -211,7 +219,11 @@ def __call__(self, z: ArrayLike, *, model_inputs: Optional[ModelInputs] = None)
             for i in range(self._resolution)
         ]
 
-    def get_activation_multiplier(self, activation: ArrayLike, *, model_inputs: Optional[ModelInputs] = None) -> ArrayLike:
```
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens/nn/distributions.py:214]: Done.

File: trulens/nn/models/keras.py (hunk 5 above)
Comment (piotrm0) [reply trulens/nn/models/keras.py:366]: Done.

File: trulens/nn/models/pytorch.py (hunk 6 above)
Comment (piotrm0) [reply trulens/nn/models/pytorch.py:71]: Generally "constant == variable" is preferable given the "constant = variable" typo becomes an error instead of a silent bug with "variable = constant" but it looks like python catches that for you so will switch.

Comment (piotrm0) [reply trulens/nn/models/pytorch.py:71]: Done.

File: trulens/nn/models/pytorch.py (hunk 7 above)
Comment (piotrm0) [reply trulens/nn/models/pytorch.py:505]: I agree but I'm saying that even if one wants one gradient, like class qoi, but implements it incorrectly (assuming class qoi was not already provided for them), then this sum might hide their mistake(s).

Comment (rshih32) [reply trulens/nn/models/pytorch.py:510]: let add a warning that this sum is usually for multi output.

File: trulens/utils/typing.py (hunk 8 above)
Comment (rshih32) [reply trulens/utils/typing.py:107]: resolved, just add a link to Lenses
maybe clarify C-> container, V->value
and that this is to help treat args/kwargs similarly as a convenience

Comment (piotrm0) [reply trulens/utils/typing.py:107]: Done.

File: tests/pytorch/unit/model_wrapper_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (piotrm0) [reply tests/pytorch/unit/model_wrapper_test.py:76]: Added explanation in other places.

File: trulens/nn/models/pytorch.py (hunk 7 above)
Comment (piotrm0) [reply trulens/nn/models/pytorch.py:510]: Added a warning only if more than 1 dimension is being aggregated.

**Review by rshih32 — APPROVED** [review]
